pymc3/gp/gp.py

```diff
@@ -2,60 +2,124 @@
-
+from ..distributions import Normal, MvNormal, Continuous, draw_values, generate_samples
+from ..model import modelcontext, Deterministic, ObservedRV
+from ..distributions.dist_math import Cholesky
 
 __all__ = ['GP', 'sample_gp']
 
+CHOL_CONST = True
+cholesky = Cholesky(nofail=True, lower=True)
+solve_lower = tt.slinalg.Solve(A_structure="lower_triangular")
+solve_upper = tt.slinalg.Solve(A_structure="upper_triangular")
+
+
+def stabilize(K):
+    if CHOL_CONST:
+        n = K.shape[0]
+        return K + 1e-6 * (tt.nlinalg.trace(K)/n) * tt.eye(n)
+    else:
+        return K
 
-class GP(Continuous):
-    """Gausian process
 
+def GP(name, X, mean_func=None, cov_func=None,
+       cov_func_noise=None, sigma=None,
+       approx=None, n_inducing=None, inducing_points=None,
+       observed=None, chol_const=True, *args, **kwargs):
+    """Gausian process constructor
     Parameters
     ----------
+    X : array
+        Grid of points to evaluate Gaussian process over.
     mean_func : Mean
         Mean function of Gaussian process
     cov_func : Covariance
         Covariance function of Gaussian process
-    X : array
-        Grid of points to evaluate Gaussian process over. Only required if the
-        GP is not an observed variable.
+    cov_func_noise : Covariance
+        Covariance function of noise process (ignored for approximations)
     sigma : scalar or array
-        Observation standard deviation (defaults to zero)
+        Noise standard deviation
+    approx : None or string,
+        Allowed values are 'FITC' and 'VFE'
+    n_inducing : integer
+        The number of inducing points to use
+    inducing_points : array
+        The inducing points to use
+    observed : None or array
+        The observed 'y' values, use if likelihood is Gaussian
+    chol_const: boolean
+        Whether or not to stabilize Cholesky decompositions
     """
-    def __init__(self, mean_func=None, cov_func=None, X=None, sigma=0, *args, **kwargs):
+    global CHOL_CONST
+    CHOL_CONST = chol_const
 
-        if mean_func is None:
-            self.M = Zero()
-        else:
-            if not isinstance(mean_func, Mean):
-                raise ValueError('mean_func must be a subclass of Mean')
-            self.M = mean_func
-
-        if cov_func is None:
-            raise ValueError('A covariance function must be specified for GPP')
-        if not isinstance(cov_func, Covariance):
-            raise ValueError('cov_func must be a subclass of Covariance')
-        self.K = cov_func
+    if mean_func is None:
+        mean_func = Zero()
+    else:
+        if not isinstance(mean_func, Mean):
+            raise ValueError('mean_func must be a subclass of Mean')
+    if cov_func is None:
+        raise ValueError('A covariance function must be specified for GPP')
+    if not isinstance(cov_func, Covariance):
+        raise ValueError('cov_func must be a subclass of Covariance')
+
+    # NONCONJUGATE
+    if observed is None:
```

Comment:
hmm.. if the gp + noise values are observed, and the noise is gaussian, it's best to use conjugacy to integrate out the actual gp so it doesnt need to be sampled.  But thinking about it more, thats only true **if** nothing happens in the model later that depends on the observed gp + noise data.   This would be kind of weird... but possible I guess.